homeassistant/components/zwave_js/triggers/event.py

```diff
@@ -0,0 +1,237 @@
+"""Offer Z-Wave JS event listening automation trigger."""
+from __future__ import annotations
+
+import functools
+from typing import Any
+
+from pydantic import ValidationError
+import voluptuous as vol
+from zwave_js_server.client import Client
+from zwave_js_server.model.controller import CONTROLLER_EVENT_MODEL_MAP
+from zwave_js_server.model.driver import DRIVER_EVENT_MODEL_MAP
+from zwave_js_server.model.node import NODE_EVENT_MODEL_MAP, Node
+
+from homeassistant.components.automation import (
+    AutomationActionType,
+    AutomationTriggerInfo,
+)
+from homeassistant.components.zwave_js.const import (
+    ATTR_CONFIG_ENTRY_ID,
+    ATTR_EVENT,
+    ATTR_EVENT_DATA,
+    ATTR_EVENT_SOURCE,
+    ATTR_NODE_ID,
+    ATTR_PARTIAL_DICT_MATCH,
+    DATA_CLIENT,
+    DOMAIN,
+)
+from homeassistant.components.zwave_js.helpers import (
+    async_get_node_from_device_id,
+    async_get_node_from_entity_id,
+    get_device_id,
+    get_home_and_node_id_from_device_entry,
+)
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.const import ATTR_DEVICE_ID, ATTR_ENTITY_ID, CONF_PLATFORM
+from homeassistant.core import CALLBACK_TYPE, HassJob, HomeAssistant, callback
+from homeassistant.helpers import config_validation as cv, device_registry as dr
+from homeassistant.helpers.typing import ConfigType
+
+# Platform type should be <DOMAIN>.<SUBMODULE_NAME>
+PLATFORM_TYPE = f"{DOMAIN}.{__name__.rsplit('.', maxsplit=1)[-1]}"
+
+EVENT_MODEL_MAP = {
+    "controller": CONTROLLER_EVENT_MODEL_MAP,
+    "driver": DRIVER_EVENT_MODEL_MAP,
+    "node": NODE_EVENT_MODEL_MAP,
+}
+
+
+def validate_non_node_event_source(obj: dict) -> dict:
+    """Validate that a trigger for a non node event source has a config entry."""
+    if obj[ATTR_EVENT_SOURCE] != "node" and ATTR_CONFIG_ENTRY_ID in obj:
+        return obj
+    raise vol.Invalid(f"Non node event triggers must contain {ATTR_CONFIG_ENTRY_ID}.")
+
+
+def validate_event_name(obj: dict) -> dict:
+    """Validate that a trigger has a valid event name."""
+    event_source = obj[ATTR_EVENT_SOURCE]
+    event_name = obj[ATTR_EVENT]
+    # the keys to the event source's model map are the event names
+    vol.In(EVENT_MODEL_MAP[event_source])(event_name)
+    return obj
+
+
+def validate_event_data(obj: dict) -> dict:
+    """Validate that a trigger has a valid event data."""
+    # Return if there's no event data to validate
+    if ATTR_EVENT_DATA not in obj:
+        return obj
+
+    event_source = obj[ATTR_EVENT_SOURCE]
+    event_name = obj[ATTR_EVENT]
+    event_data = obj[ATTR_EVENT_DATA]
+    try:
+        EVENT_MODEL_MAP[event_source][event_name](**event_data)
+    except ValidationError as exc:
+        errors: list[dict[str, Any]] = []
+        for error in exc.errors():
+            # Filter out required field errors if keys can be missing
+            if error["msg"] == "field required":
+                continue
+            errors.append(error)  # type: ignore
```

Comment:
Please add the error code in words to the ignore comment inside brackets.

Reply:
I changed the approach so this is no longer necessary